cep-XXXX.md

```diff
@@ -0,0 +1,237 @@
+
+### Allowed Characters and Formats for conda Channels, Subdirs, Labels, and Package Names
+
+The follow regexes define valid conda channel names, labels, and conda package names for conda packages that can be stored in an OCI registry.
+
+- channel: `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`
+- subdirs: `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`
+- label: `^[a-zA-Z][0-9a-zA-Z_\-\.\/:\s]*`
+- package name: `^(([a-z0-9])|([a-z0-9_](?!_)))[._-]?([a-z0-9]+(\.|-|_|$))*$`
+
+All channels, subdirs, labels, and package names MUST conform to their respective regex in the list above if the package is to be stored in an OCI registry.
+
+Further, the following rule applies to labels:
+
+- The label `NOLABEL` is reserved and MUST only be used for conda packages which have no other labels. In other words, in the space of labels, the empty set is represented by the label `NOLABEL`.
+
+### Mapping conda Packages with Channels, Subdirs, Labels to OCI Repositories and Tags
+
+Per the [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec) v1, the `<name>` and `<tag>` of an OCI repository are subject to the following regexes:
+
+- `<name>`: `^[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*(\/[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*)*$`
+- `<tag>`: `^[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}$`
+
+The [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec) further notes that "Many clients impose a limit of 255 characters on the length of the concatenation of the registry hostname (and optional port), /, and <name> value."
+
+These constraints motivate the following set of rules for mapping conda packages to OCI repositories and tags. More explanation is given below in the Rationale section and the example implementation illustrates how these rules work together to create a human-readable and easily parsable OCI repository `<name>` and `<tag>` for a conda package.
+
+#### Overall Form of the OCI Repository and Tag
+
+A conda package with its channel, subdir, and (optional) label MUST be mapped to an OCI `<repository>:<tag>` in one of two forms.
+
+The unhashed form of the OCI artifact `<name>:<tag>` for a conda package is:
+
+```
+<channel>/<subdir>/<OCI-encoded package name>:<OCI-encoded version>-<OCI-encoded build>(-<OCI-encoded label>)
+```
+
+where the `-<OCI-encoded label>` component MUST be present if the package label is not `main`.
+
+In the hashed form, the OCI-form of the package name and the OCI tag (i.e., `<OCI-encoded version>-<OCI-encoded build>(-<OCI-encoded label>)`) are each separately hashed via SHA1 and then replaced by the string `h<hexdigest of SHA1 hash>` as follows:
+
+```
+<channel>/<subdir>/h<hexdigest of the SHA1 hash of OCI-encoded package name>:h<hexdigest of the SHA1 hash of the original OCI tag of the conda package>
+```
+
+The `<channel>` and `<subdir>` MUST be used as-is without modification. The OCI-forms of the package name, label, version, and build string are defined below.
```

Comment:
What if the channel name along is too long to be encoded? Would it be ok to have top-level `h<hash>` entries with the corresponding `org.conda.channel.{name,subdir,label}` annotation metadata?

Reply:
Yes maybe just doing that all the time is simpler actually. I was trying to increase readability, but the complexity cost is high.